File: solutions/ocaml/157-middle-of-the-linked-list.ml
(*
  Problem 157: Middle of the Linked List (LeetCode 876)
  Difficulty: Easy
  Language: OCaml
*)
type list_node = { v: int; next: list_node option }

let middle_node head =
  let rec collect = function
    | None -> []
    | Some n -> n :: collect n.next in
  let nodes = collect head in
  let mid = List.length nodes / 2 in
  List.nth_opt nodes mid

File: solutions/scala/157-middle-of-the-linked-list.scala
/**
 * Problem 157: Middle of the Linked List (LeetCode 876)
 * Difficulty: Easy
 * Language: Scala
 */
case class ListNode(v: Int, next: Option[ListNode])
def middleNode(head: Option[ListNode]): Option[ListNode] = {
  def collect(n: Option[ListNode]): List[ListNode] = n match {
    case None => Nil
    case Some(node) => node :: collect(node.next)
  }
  val nodes = collect(head)
  Some(nodes(nodes.length / 2))
}
